Source/WebCore/rendering/RenderGrid.cpp

```diff
@@ -2293,62 +2251,81 @@ void RenderGrid::computeContentPositionAndDistributionOffset(GridTrackSizingDire
-    if (isRowAxis ? isSubgridColumns() : isSubgridRows()) {
-        offset.positionOffset = 0_lu;
-        offset.distributionOffset = 0_lu;
+
+    // Initialize to start alignment.
+    offset.positionOffset = 0_lu;
+    offset.distributionOffset = 0_lu;
+
+    if (isRowAxis ? isSubgridColumns() : isSubgridRows())
         return;
-    }
+
     auto contentAlignmentData = contentAlignment(direction);
+    auto distribution = contentAlignmentData.distribution();
     auto position = contentAlignmentData.position();
-    // If <content-distribution> value can't be applied, 'position' will become the associated
-    // <content-position> fallback value.
-    contentDistributionOffset(offset, availableFreeSpace, position, contentAlignmentData.distribution(), numberOfGridTracks);
-    if (offset.isValid())
-        return;
 
-    if (availableFreeSpace <= 0 && contentAlignmentData.overflow() == OverflowAlignment::Safe) {
-        offset.positionOffset = 0_lu;
-        offset.distributionOffset = 0_lu;
-        return;
+    // Apply <content-distribution> and return, or continue to fallback positioning if we can't distribute.
+    if (distribution != ContentDistribution::Default) {
+        if (position == ContentPosition::Normal)
+            position = resolveContentDistributionFallback(distribution);
+        if (availableFreeSpace > 0) {
+            switch (distribution) {
+            case ContentDistribution::SpaceBetween:
+                if (numberOfGridTracks < 2)
+                    break;
+                offset.distributionOffset = availableFreeSpace / (numberOfGridTracks - 1);
+                return;
+            case ContentDistribution::SpaceAround:
+                if (numberOfGridTracks < 1)
+                    break;
+                offset.distributionOffset = availableFreeSpace / numberOfGridTracks;
+                offset.positionOffset = offset.distributionOffset / 2;
+                return;
+            case ContentDistribution::SpaceEvenly:
+                offset.distributionOffset = availableFreeSpace / (numberOfGridTracks + 1);
+                offset.positionOffset = offset.distributionOffset;
+                return;
+            case ContentDistribution::Stretch:
+                break;
+            default:
+                ASSERT_NOT_REACHED();
+            }
+        }
     }
 
-    LayoutUnit positionOffset;
+    // Apply alignment safety.
+    if (availableFreeSpace <= 0 && contentAlignmentData.overflow() == OverflowAlignment::Safe)
+        return;
+
+    // Apply <content-position> / fallback positioning.
     switch (position) {
     case ContentPosition::Left:
         ASSERT(isRowAxis);
-        positionOffset = style().isLeftToRightDirection() ? 0_lu : availableFreeSpace;
-        break;
+        if (!style().isLeftToRightDirection())
+            offset.positionOffset = availableFreeSpace;
+        return;
     case ContentPosition::Right:
-        ASSERT(isRowAxis);
```

Comment:
Is removing this assert ok? AFAICT the spec says that left and right work with the justify properties and I see we kept it in the `ContentPosition::Left` case

Reply:
Agree that's a mistake, fixed.